### DISCUSSION TITLE
How does one use a RestClient within an Extension?

### DISCUSSION THREAD

**Discussion by akil-rails** [quarkusio/quarkus#34180] [Q&A] [answered]
I need to create a Quarkus extension that includes a rest client - registered via @RegisterRestClient. I am following the tutorial in the [docs](https://quarkus.io/guides/building-my-first-extension)

The rest client:
```
@RegisterRestClient(configKey = "rta-cams")
@RegisterClientHeaders(TokenClientHeader.class) 
@Path("restAuth/api/v1/getToken")
public interface TokenClient {
  @POST
  Uni<TokenResponse> getToken(TokenRequest request);
}
```

The buildstep
```
    @BuildStep
    AdditionalBeanBuildItem createTokenService() {
        return AdditionalBeanBuildItem                    
            .builder()
            .addBeanClass(TokenService.class)
            .addBeanClass(TokenMapper.class)
            .addBeanClass(TokenClient.class)
            .build();
    }
```

When I include the extension in the app, the build fails with the following error
```
Caused by: jakarta.enterprise.inject.UnsatisfiedResolutionException: Unsatisfied dependency for type com.apibanking.mf.services.core.runtime.infrastructure.remote.clients.token.TokenClient and qualifiers [@RestClient]
	- java member: com.apibanking.mf.services.core.runtime.application.services.TokenService#tokenClient
	- declared on CLASS bean [types=[java.lang.Object, com.apibanking.mf.services.core.runtime.application.services.TokenService], qualifiers=[@Default, @Any], target=com.apibanking.mf.services.core.runtime.application.services.TokenService]
```

The TokenService and TokenMapper are other classes, which are correctly discovered during the build of the app, but not the client class. What do i need to add to make sure that the RegisterRestClient annotations are correctly handled?

**Comment by akil-rails** [comment]
I tried using a custom builder, using QuarkusRestClientBuilder, but then the error shifts to the following
```
java.lang.IllegalArgumentException: Not a REST client interface: interface com.apibanking.mf.services.core.runtime.infrastructure.remote.clients.token.TokenRestClient. No @Path annotation found on the class or any methods of the interface and no HTTP method annotations (@POST, @PUT, @GET, @HEAD, @DELETE, etc) found on any of the methods
```

Struggling to figure out how to get a Rest Client work within a quarkus extension - would really appreciate any help!

**Comment by Pilpin** [comment]
Hello, I have the same question, did you find a way to make it work ?

**Reply by akil-rails** [reply]
Yes - pls see https://github.com/quarkusio/quarkus/issues/34267

I added an empty beans.xml in the runtime module, and it worked

**Reply by dcdh** [reply]
You should not add a beans.xml inside the runtime module because all beans presents in the runtime module will be indexed even the one you may not want. Thus, it make the AdditionnalBeanBuildItem useless.

**Comment by mschorsch** [comment, marked as answer]
Use `AdditionalIndexedClassesBuildItem` instead of `AdditionalBeanBuildItem`.

```java
    @BuildStep
    AdditionalIndexedClassesBuildItem addAdditionalClasses() {
        return AdditionalIndexedClassesBuildItem(TokenClient.class.getName());
    }
```
